### ISSUE TITLE
from_sql and to_sql would be really helpful features in vaex

### ISSUE THREAD

**Issue by vishalhkd007** [vaexio/vaex#864]
Vaex has a huge advantage over Pandas when it comes to loading data from files into dataframe. With support for data read and write to RDBMS, Vaex can become a powerful tool for ETL processing where we currently are using Pandas (which is slow for large datasets).

The advantage of Pandas is its support for the database systems. I can create a dataframe from an SQL query as well as stream data to an RDBMS that creates a database table from the Pandas dataframe. This is quite helpful when it comes to using Pandas for ETL where processing of data including dealing with missing data, flagging incorrect data etc is done in Pandas and then streamed to database.

Vaex for its performance gains for large dataset over Pandas would really gain from basic database support making it the tool for such processing.

**Comment by maartenbreddels** [comment]
Hi,

 I was thinking about it. And I'm open to the idea, possible via https://turbodbc.readthedocs.io/en/latest/ ->arrow -> vaex and possibly via arrow in the future directly.
What kind of usage do you see, a batch downloading to a arrow/hdf5 file after which you work with the data, or a 'lazy' fetch from the database, where no data is kept in memory?

cheers,

Maarten

**Comment by vishalhkd007** [comment]
Thanks for replying on this issue.



The usage mentioned is one that is quite common for data transfers and ETL processing from legacy systems to RDBMS or NoSQL systems. We deal with a lot of clients who are trying to have this kind of data migration where data is dumped to CSV files and from there, the data needs to be processed and analyzed for any discrepancies and missing data. 



For small datasets, this can be done with Pandas. But our clients have data in sizes over 100GB+. In these situations, loading data in-memory for analysis is not an option. One way to deal with these is to split the data into smaller files (but many operations that we do on the data do not work with such splits). Vaex shows a lot of potential here where the lazy loading from arrow/hdf5 files works and allows for the ETL analysis.



Another use case is data consolidation - where we are dealing with data acquired from different sources (either through mergers or where our clients bring in external data sources to be mashed with what they already have). Now, we are dealing with multiple datasets which need to be analyzed for commonality along with the need for joins and merges. Again, based on the size of the data, this can be an extremely slow and painful in Pandas. I am hopeful that Vaex will be beneficial - in fact, I have been successful in creating a mash of over 5M records from 2 datasets using Vaex where it outshined Pandas by a huge margin.



Once merged, I need to stream this to a RDBMS table. Here, I am forced to convert the vaex dataframe to Pandas and then stream that to the table. I would prefer to have this ability from vaex itself.



Hope this helps ...
 
Vishal Deshpande

Founder, MTekLabs

vishal@mteklabs.com​

http://www.mteklabs.com​

+1 - 540 454 8175






---- On Fri, 26 Jun 2020 08:37:31 -0500 Maarten Breddels <notifications@github.com> wrote ----





Hi,

I was thinking about it. And I'm open to the idea, possible via https://turbodbc.readthedocs.io/en/latest/ ->arrow -> vaex and possibly via arrow in the future directly.
 What kind of usage do you see, a batch downloading to a arrow/hdf5 file after which you work with the data, or a 'lazy' fetch from the database, where no data is kept in memory?
cheers,

Maarten

**Comment by alexrabe91** [comment]
> Hi,
> 
> I was thinking about it. And I'm open to the idea, possible via https://turbodbc.readthedocs.io/en/latest/ ->arrow -> vaex and possibly via arrow in the future directly.
> What kind of usage do you see, a batch downloading to a arrow/hdf5 file after which you work with the data, or a 'lazy' fetch from the database, where no data is kept in memory?
> 
> cheers,
> 
> Maarten

just my 2 cents to this conversation: I would also love to see an integrated "from/to_sql" feature. We often have to deal with huge csv files (in a range from a few GB to hundreds of GB) coming as dump directly from a database where the current approach with vaex requires a transformation from csv to arrow/hdf5 in order to utilize vaex. I would love to see a batch download as well as a lazy fetch since I can directly imagine different scenarios for our side where both could take place.

cheers,
alex

**Comment by maartenbreddels** [comment]
Yes, I think skipping the CSV part will be highly beneficial, but can already be done using iterator/chunks:
https://docs.vaex.io/en/latest/example_io.html
Or looking at
https://github.com/vaexio/vaex/blob/8d3a0249a7ca6b75bf6bf120fb147fc9a651f928/packages/vaex-core/vaex/dataframe.py#L5890

I think all the building blocks are already in place, we can make this more user friendly though, and it would require some more testing/benchmarking to make this really solid. We're open to the idea of external funding to develop this.
